ledger/accountdb.go

```diff
@@ -2706,52 +2678,77 @@ func (iterator *encodedAccountsBatchIter) Next(ctx context.Context, tx *sql.Tx,
 		}
 	}
 	if iterator.resourcesRows == nil {
-		iterator.resourcesRows, err = tx.QueryContext(ctx, "SELECT addrid, aidx, data FROM resources ORDER BY addrid, aidx, rtype")
+		iterator.resourcesRows, err = tx.QueryContext(ctx, "SELECT addrid, aidx, rtype, data FROM resources ORDER BY addrid, aidx, rtype")
 		if err != nil {
 			return
 		}
 	}
 
 	// gather up to accountCount encoded accounts.
 	bals = make([]encodedBalanceRecordV6, 0, accountCount)
-	var addr basics.Address
-	for iterator.accountsRows.Next() {
-		var addrbuf []byte
-		var buf []byte
-		var rowid int64
-		err = iterator.accountsRows.Scan(&rowid, &addrbuf, &buf)
-		if err != nil {
-			iterator.Close()
-			return
-		}
-
-		if len(addrbuf) != len(addr) {
-			err = fmt.Errorf("account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
-			return
-		}
+	var encodedRecord encodedBalanceRecordV6
+	var baseAcct baseAccountData
+	var numAcct int
+	baseCb := func(addr basics.Address, rowid int64, accountData *baseAccountData, encodedAccountData []byte) (err error) {
+		encodedRecord = encodedBalanceRecordV6{Address: addr, AccountData: encodedAccountData}
+		baseAcct = *accountData
+		numAcct++
+		return nil
+	}
 
-		copy(addr[:], addrbuf)
-		encodedRecord := encodedBalanceRecordV6{Address: addr, AccountData: buf}
-		for {
-			var nextRes *endcodedResourcesRow
-			nextRes, err = iterator.peekNextResourcesRow()
-			if err != nil {
-				return
-			}
-			if nextRes.addrid != rowid {
-				break
-			}
+	var totalAppParams, totalAppLocalStates, totalAssetParams, totalAssets uint32
+	// emptyCount := 0
+	resCb := func(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType, resData *resourcesData, encodedResourceData []byte) error {
+		emptyBaseAcct := baseAcct.TotalAppParams == 0 && baseAcct.TotalAppLocalStates == 0 && baseAcct.TotalAssetParams == 0 && baseAcct.TotalAssets == 0
+		if !emptyBaseAcct && resData != nil {
```

Comment:
`emptyBaseAcct` means a current or prev account (relative to what was just read from resources table) does not have any resources.
`resData` is null when no resource at this call.

If I recall it right this condition handles a case when a very first account has 0 resources and the second has some.